[ResourceMonitor] Track network usage of SharedWorker.

basuke opened the pull request:
#### deae73529c4c33d969518fd587cfac3f308d74c9
<pre>
[ResourceMonitor] Track network usage of SharedWorker.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287464">https://bugs.webkit.org/show_bug.cgi?id=287464</a>
<a href="https://rdar.apple.com/142654034">rdar://142654034</a>

Reviewed by Ben Nham.

We don&apos;t track network usage of SharedWorker created by iframe and also there&apos;s no direct connection
between SharedWorker and iframe right now. To monitor network usage, add new IPC message to report
network usage of SharedWorker to all connected iframe document.

To make this happen:
1. Add new parameter, `workerIdentifier` in NetworkResourceLoadParameters.
2. Report network usage of the SharedWorker from NetworkResourceLoader to connected SharedWorker objects.
3. Tell ResourceMonitor to treate those usage as part of the resource usage of iframe.

New layout test is added to test this usage.

* LayoutTests/http/tests/iframe-monitor/blob-resource.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/cached-resource.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/dark-mode.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/data-url-resource.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/eligibility.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/iframe-unload.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/shared-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/shared-worker.js: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js: Refactoring
* LayoutTests/http/tests/iframe-monitor/shared-worker-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/shared-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/throttler.html: Refactoring
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/WorkerThreadableLoader.cpp:
(WebCore::m_contextIdentifier):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::m_resourceMonitor):
(WebCore::SharedWorker::reportNetworkUsage):
* Source/WebCore/workers/shared/SharedWorker.h:
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.cpp:
(WebCore::SharedWorkerObjectConnection::reportNetworkUsageToWorkerObject):
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::bufferingTimerFired):
(WebKit::NetworkResourceLoader::sendBuffer):
(WebKit::NetworkResourceLoader::dataReceivedThroughContentFilter):
(WebKit::NetworkResourceLoader::sendDidReceiveDataMessage):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllSharedWorkers):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::reportNetworkUsageToAllSharedWorkerClients):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.cpp:
(WebKit::WebSharedWorkerServerConnection::reportNetworkUsageToWorkerObject):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerObjectConnection.messages.in:

Canonical link: <a href="https://commits.webkit.org/290232@main">https://commits.webkit.org/290232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66edd0beb254ac946f9d635da5c3adb0ea9ae21a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89338 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8863 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44177 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94324 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40099 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91389 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9250 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17109 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68835 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26499 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92340 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7098 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81085 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49196 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6844 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35470 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39206 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77200 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96153 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16518 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77706 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16774 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76876 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77007 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19002 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21428 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20029 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9668 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16532 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21843 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16273 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19724 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebCore/workers/shared/SharedWorkerObjectConnection.cpp:
```diff
@@ -94,6 +94,15 @@ void SharedWorkerObjectConnection::postErrorToWorkerObject(SharedWorkerObjectIde
+{
+    ASSERT(isMainThread());
+    auto* workerObject = SharedWorker::fromIdentifier(sharedWorkerObjectIdentifier);
+    CONNECTION_RELEASE_LOG_ERROR("sendNetworkUsageToWorkerObject: sharedWorkerObjectIdentifier=%" PUBLIC_LOG_STRING ", worker=%p", sharedWorkerObjectIdentifier.toString().utf8().data(), workerObject);
```
Seems like this should not be `RELEASE_LOG_ERROR`. I would suggest that it should either be a regular `RELEASE_LOG` that also logs `bytesTransferredOverNetworkDelta`, or should log nothing at all.

bnham commented on Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
```diff
@@ -68,6 +68,7 @@ NetworkResourceLoadParameters::NetworkResourceLoadParameters(
 #if ENABLE(CONTENT_EXTENSIONS)
     , URL&& mainDocumentURL
     , std::optional<UserContentControllerIdentifier> userContentControllerIdentifier
+    , std::variant<std::monostate, WebCore::SharedWorkerIdentifier> workerIdentifier
```
nit: seems like `std::optional` is more appropriate here.

bnham commented on Source/WebCore/loader/ResourceLoaderOptions.h:
```diff
@@ -255,6 +257,7 @@ struct ResourceLoaderOptions : public FetchOptions {
 
     Markable<FetchIdentifier> navigationPreloadIdentifier;
     String nonce;
+    std::variant<std::monostate, SharedWorkerIdentifier> workerIdentifier;
```
Seems like `std::optional` is more appropriate here.

bnham commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2181,6 +2182,32 @@ void NetworkResourceLoader::useRedirectionForCurrentNavigation(WebCore::Resource
+    UNUSED_PARAM(delta);
+    Ref connection = m_connection;
+    if (auto* session = connection->protectedNetworkProcess()->networkSession(sessionID())) {
+        if (auto* server = session->sharedWorkerServer())
```
I think the bot will want this to be `RefPtr server = session->sharedWorkerServer()`

bnham commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2181,6 +2182,32 @@ void NetworkResourceLoader::useRedirectionForCurrentNavigation(WebCore::Resource
+{
+    UNUSED_PARAM(delta);
+    Ref connection = m_connection;
+    if (auto* session = connection->protectedNetworkProcess()->networkSession(sessionID())) {
```
I think the safer C++ bot will want this to be `RefPtr session = ...`

basuke replied:
NetworkSession is not RefCounted but CanMakeCheckedPtr so I will use `CheckedPtr` here.

bnham commented on Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.cpp:
```diff
@@ -146,6 +146,12 @@ void WebSharedWorkerServerConnection::postErrorToWorkerObject(WebCore::SharedWor
 
+void WebSharedWorkerServerConnection::reportNetworkUsageToWorkerObject(WebCore::SharedWorkerObjectIdentifier sharedWorkerObjectIdentifier, size_t bytesTransferredOverNetworkDelta)
+{
+    CONNECTION_RELEASE_LOG_ERROR("reportNetworkUsageToWorkerObject: sharedWorkerObjectIdentifier=%" PUBLIC_LOG_STRING, sharedWorkerObjectIdentifier.toString().utf8().data());
```
This should either be a `RELEASE_LOG` (not `RELEASE_LOG_ERROR`) or not be logged at all.

basuke committed:
[ResourceMonitor] Track network usage of SharedWorker.
https://bugs.webkit.org/show_bug.cgi?id=287464
rdar://142654034

Reviewed by Ben Nham.

We don't track network usage of SharedWorker created by iframe and also there's no direct connection
between SharedWorker and iframe right now. To monitor network usage, add new IPC message to report
network usage of SharedWorker to all connected iframe document.

To make this happen:
1. Add new parameter, `workerIdentifier` in NetworkResourceLoadParameters.
2. Report network usage of the SharedWorker from NetworkResourceLoader to connected SharedWorker objects.
3. Tell ResourceMonitor to treate those usage as part of the resource usage of iframe.

New layout test is added to test this usage.

* LayoutTests/http/tests/iframe-monitor/blob-resource.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/cached-resource.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/dark-mode.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/data-url-resource.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/eligibility.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/iframe-unload.html: Refactoring
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/shared-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/shared-worker.js: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js: Refactoring
* LayoutTests/http/tests/iframe-monitor/shared-worker-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/shared-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/throttler.html: Refactoring
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/WorkerThreadableLoader.cpp:
(WebCore::m_contextIdentifier):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::m_resourceMonitor):
(WebCore::SharedWorker::reportNetworkUsage):
* Source/WebCore/workers/shared/SharedWorker.h:
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.cpp:
(WebCore::SharedWorkerObjectConnection::reportNetworkUsageToWorkerObject):
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::bufferingTimerFired):
(WebKit::NetworkResourceLoader::sendBuffer):
(WebKit::NetworkResourceLoader::dataReceivedThroughContentFilter):
(WebKit::NetworkResourceLoader::sendDidReceiveDataMessage):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllSharedWorkers):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::reportNetworkUsageToAllSharedWorkerClients):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.cpp:
(WebKit::WebSharedWorkerServerConnection::reportNetworkUsageToWorkerObject):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerObjectConnection.messages.in:

Canonical link: https://commits.webkit.org/290232@main
